lib/cdo/google/sheet.rb

```diff
@@ -21,29 +21,52 @@ def initialize(document_key)
       @document_key = document_key
     end
 
-    def export(sheet_name:, rows:)
+    def export(tab_name:, data:)
       # Write exported data to a sheet in the document
-      @drive.update_sheet rows, @document_key, "#{sheet_name} (auto)"
+      @drive.update_sheet data, @document_key, "#{tab_name} (auto)"
 
       # Write new metadata to a second sheet in the document
       last_updated = DateTime.now.in_time_zone ActiveSupport::TimeZone.new "Pacific Time (US & Canada)"
       metadata = <<~META.split("\n").map {|line| [line]}
         AUTOMATION METADATA
 
-        The tabs "#{sheet_name} (auto)" and "#{sheet_name}_meta (auto)" are auto-generated;
+        The tabs "#{tab_name} (auto)" and "#{tab_name}_meta (auto)" are auto-generated;
         Any edits you make to them (besides formatting) may be lost.
 
         Last updated: #{last_updated.strftime '%Y-%m-%d %l:%M%P GMT%:::z'}
         Written by: #{CDO.gdrive_export_secret.client_email}
-        Rows: #{rows.length - 1}
+        Rows: #{data.length - 1}
 
         Parts of this Google Sheet are auto-populated from our live application by an automated process.
         The sheet is shared with a \"service account\" that updates it on the application's behalf.
         (Technical Details: https://github.com/code-dot-org/code-dot-org/pull/32597)
       META
-      @drive.update_sheet metadata, @document_key, "#{sheet_name}_meta (auto)"
+      @drive.update_sheet metadata, @document_key, "#{tab_name}_meta (auto)"
     end
 
+    def notify_of_external_sharing(dcdo_key=nil)
+      # List of external emails that we can share this document with.
+      # This list is saved in DCDO as an array. To append a new value to this list:
+      # DCDO.set(key_name, DCDO.get(key_name, []) << new_value)
+      allowed_list = dcdo_key.nil? ? [] : DCDO.get(dcdo_key, [])
+      external_emails = external_emails_with_access - allowed_list
+
+      if external_emails.present?
+        email_domains = external_emails.map {|email| email.slice(/@.*/)}.uniq
+        error_msg = <<~ERROR_MSG
+          Document containing PII information is shared to
+          #{external_emails.length} external account(s) at the following domain(s): #{email_domains.join(', ')}.
+          Document key: #{@document_key}
```

Comment:
This...may not be a great idea. Basically I'm putting something that is in a secret in an error message -- other ideas to identify where this error has occurred?

Reply:
hmm, yeah I agree that this seems potentially undesirable. I'd suggest adding an optional parameter the calling methods can pass in to act as an identifier. I _think_ that even in the absence of an identifier, the Honeybadger trace should be sufficient to identify from where this is being called, so making the param optional shouldn't break anything.